From 7cf4b9778f8114e1da7f58f5ac3f08f44bc43f56 Mon Sep 17 00:00:00 2001
From: Michael-CY Lee <michael-cy.lee@mediatek.com>
Date: Tue, 5 Aug 2025 10:20:17 +0800
Subject: [PATCH 227/236] mtk: hostapd: always do WPA initialization in
 hostapd_reload_bss() if hapd->conf->wpa is set

Compared to the WPA initialization flow, a lot of content is missing
in the WPA reconfiguration flow. For example, "tx_bss_auth" and
"tx_status" on struct wpa_auth_config are not considered to be set in
the WPA reconfiguration flow.

To keep the content consistent, this change always uses the initialization
flow whenever hapd->conf->wpa is set.

Signed-off-by: Michael-CY Lee <michael-cy.lee@mediatek.com>
---
 src/ap/hostapd.c | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/src/ap/hostapd.c b/src/ap/hostapd.c
index 22b1b80ff..fb5d7b687 100644
--- a/src/ap/hostapd.c
+++ b/src/ap/hostapd.c
@@ -163,10 +163,14 @@ static void hostapd_reload_bss(struct hostapd_data *hapd)
 	else
 		hostapd_set_drv_ieee8021x(hapd, hapd->conf->iface, 0);
 
-	if (hapd->conf->wpa && hapd->wpa_auth == NULL) {
+	if (hapd->conf->wpa) {
 		hostapd_setup_wpa(hapd);
 		if (hapd->wpa_auth)
 			wpa_init_keys(hapd->wpa_auth);
+	/* There are a lot of missing content in the WPA reconfiguration flow
+	 * compared to the WPA initialization flow, so deprecate it.
+	 */
+	/*
 	} else if (hapd->conf->wpa) {
 		const u8 *wpa_ie;
 		size_t wpa_ie_len;
@@ -175,6 +179,7 @@ static void hostapd_reload_bss(struct hostapd_data *hapd)
 		if (hostapd_set_generic_elem(hapd, wpa_ie, wpa_ie_len))
 			wpa_printf(MSG_ERROR, "Failed to configure WPA IE for "
 				   "the kernel driver.");
+	*/
 	} else if (hapd->wpa_auth) {
 		wpa_deinit(hapd->wpa_auth);
 		hapd->wpa_auth = NULL;
-- 
2.45.2

